example/contrib/socks-client/socks/detail/char_type.hpp

```diff
@@ -0,0 +1,193 @@
+//
+// Copyright (c) 2018 jackarain (jack dot wgm at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef SOCKS_DETAIL_CHAR_TYPE_HPP
+#define SOCKS_DETAIL_CHAR_TYPE_HPP
+
+#include <socks/config.hpp>
+#include <string>
+#include <stdexcept>
+#include <iterator>
+
+namespace socks {
+namespace detail {
+
+inline
+bool
+isdigit(const char c) noexcept
+{
+    return c >= '0' && c <= '9';
+}
+
+inline
+bool
+isalpha(const char c) noexcept
+{
+    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
+}
+
+inline
+bool
+ishexdigit(const char c) noexcept
+{
+    return isdigit(c) || ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z'));
+}
+
+inline
+bool
+isunreserved(const char c) noexcept
+{
+    if (isalpha(c) || isdigit(c) ||
+        c == '-' || c == '.' ||
+        c == '_' || c == '~')
+        return true;
+    return false;
+}
+
+inline
+bool
+isuchar(const char c) noexcept
+{
+    if (isunreserved(c) ||
+        c == ';' || c == '?' || c == '&' || c == '=')
+        return true;
+    return false;
+}
+
+inline
+bool
+ishsegment(const char c) noexcept
+{
+    if (isuchar(c) ||
+        c == ';' || c == ':' ||
+        c == '@' || c == '&' || c == '=')
+        return true;
+    return false;
+}
+
+inline
+bool
+issubdelims(const char c) noexcept
+{
+    if (c == '!' || c == '$' || c == '\'' || c == '(' ||
+        c == ')' || c == '*' || c == '+'  || c == ',' ||
+        c == '=')
+        return true;
+    return false;
+}
+
+inline
+bool
+uri_reserved(const char c) noexcept
+{
+    if (c == ';' || c == '/' || c == '?' || c == ':' ||
+        c == '@' || c == '&' || c == '=' || c == '+' ||
+        c == '$' || c == ',')
+        return true;
+    return false;
+}
+
+inline
+bool
+uri_mark(const char c) noexcept
+{
+    if (c == '-' || c == '_' || c == '.'  || c == '!' ||
+        c == '~' || c == '*' || c == '\'' || c == '(' ||
+        c == ')')
+        return true;
+    return false;
+}
+
+inline
+string_view
+to_hex(unsigned char c) noexcept
+{
+    static const char* hexstring[] = {
```

Comment:
This table could be made 1/3rd smaller by not inserting nulls, e.g.:
```
static char const hex[] = "00" "01" "02" "03" ...
```
(Note how there are no commas). I'm not sure I care enough about it to change it but I just wanted to point it out. If and when we make this a public API then yeah I would change it.

Reply:
label: comment-1

Reply:
Or `const char hexstring[][2] = {` with commas

Reply:
b08272525 (HEAD -> socks-example, origin/socks-example) comment-1 : detail::to_hex now implemented in terms of continuous string. Added a test